### ISSUE TITLE
play.py: Joystick not responding since installing gym-mupen64plus

### ISSUE THREAD

**Issue by xaviervasques** [bzier/gym-mupen64plus#19]
Hello, I think I am in the final step :)
I installed mupen64plus, my joystick, everything went smoothly. I run record.py, joystick inputs responding. I run play.py, I have the two screens opened but when the race starts, nothing, mario do not move. And I realized that my joystick in no more responding since I followed the installation of gym-mupen64plus.
If I start from scratch mupen64plus's install, it works. Then, when I install gym-mupen64plus, joystick doesn't work. The joystick is still recognized and get the right signals in record.py, when I test it through jstest, it is ok etc ... I think there is something modifying mupen64plus
Can you please help me ? I am in the final step :))

**Comment by kevinhughes27** [comment]
The data in the final step is not supposed to come from your joystick anymore it comes from the AI that was trained. Its possible the AI isn't getting the right signal from play.py. That is what I would recommend debugging.

**Comment by xaviervasques** [comment]
Ok Kevin, I will check in play.py what is happening. If somenone as an idea do not hesitate :)

**Comment by xaviervasques** [comment]
Do you think is it normal when I print joystick in play.py:

        if not manual_override:
            ## Look
            vec = resize_image(obs)
            vec = np.expand_dims(vec, axis=0) 
            ## Think
            joystick = self.model.predict(vec, batch_size=1)[0]
	    print joystick

I get numbers, but always the same as you can see in the extract:

beginning episode loop
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05
  -1.63973527e-05]
AI: [0, 0, 0, 0, 0]
[  3.93490028e-03   3.79063748e-03   4.80525123e-05  -4.07045482e-05

Output is giving always 0 .... And play.py seems to send the right signals as it goes through the menus selecting the pilot (Mario) and the race ...

Would not be the output of the train.py ?

**Comment by emomicrowave** [comment]
The values in the samples seem way too small, and probably because of that the model didn't train correctly.

Try printing the joystick input from `record.py` and when you start `record.py`, see if the lines on the graph go all the way to *1* or *-1*

**Comment by xaviervasques** [comment]
Hello and thanks for your reply. Yes, record.py print the joystick input with the colors. Acceleration is ok now with play.py but not the directions (right and left). 

An extract of the inputs received from record.py : 

(106, ' ', array([ 0.00753784,  0.00167847,  1.        ,  0.        ,  0.        ]))
(107, ' ', array([ 0.0057373 ,  0.00192261,  1.        ,  0.        ,  0.        ]))
(108, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(109, ' ', array([ 0.00769043,  0.00265503,  1.        ,  0.        ,  0.        ]))
(110, ' ', array([ 0.00778198,  0.00222778,  1.        ,  0.        ,  0.        ]))
(111, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(112, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(113, ' ', array([ 0.00408936,  0.00357056,  1.        ,  0.        ,  0.        ]))
(114, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(115, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(116, ' ', array([ 0.        ,  0.00378418,  1.        ,  0.        ,  0.        ]))
(117, ' ', array([ 0.00387573,  0.00378418,  1.        ,  0.        ,  0.        ]))
(118, ' ', array([ 0.        ,  0.00366211,  1.        ,  0.        ,  0.        ]))
(119, ' ', array([ 0.00384521,  0.00375366,  1.        ,  0.        ,  0.        ]))
(120, ' ', array([ 0.00415039,  0.00378418,  1.        ,  0.        ,  0.        ]))
(121, ' ', array([ 0.00384521,  0.00378418,  1.        ,  0.        ,  0.        ]))
(122, ' ', array([ 0.00384521,  0.00378418,  1.        ,  0.        ,  0.        ]))
(123, ' ', array([ 0.00384521,  0.00378418,  1.        ,  0.        ,  0.        ]))
(124, ' ', array([ 0.00396729,  0.00369263,  1.        ,  0.        ,  0.        ]))
(125, ' ', array([ 0.00778198,  0.00265503,  1.        ,  0.        ,  0.        ]))
(126, ' ', array([ 0.00778198,  0.00238037,  1.        ,  0.        ,  0.        ]))

**Comment by kevinhughes27** [comment]
The menu is hard coded so that is why it works and the driving does not. What did your training set look like?

**Comment by xaviervasques** [comment]
Hello, I tried few things like working on the weights of buttons (multiplying by a couple of values) because at the beginning I though it was due to my PS4 controller. So I decided to use a XBOX controller and nothing really change. Mario is going straight ahead :)
I trained TensorKart with 5 Mario Race, 5 Luigi Race, 2 Kalimari and 5 Royal Raceway. 

Here how it looks like with Xbox xontroller (record.py):
(82, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(83, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(84, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(85, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(86, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(87, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(88, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(89, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(90, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(91, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(92, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(93, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(94, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(95, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(96, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(97, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(98, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(99, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(100, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(101, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(102, ' ', array([ -1.00000000e+00,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(103, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(104, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(105, ' ', array([ -2.96875000e-01,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(106, ' ', array([ -2.10937500e-01,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(107, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(108, ' ', array([ -9.84375000e-01,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(109, ' ', array([ -1.00000000e+00,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(110, ' ', array([  3.12500000e-02,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(111, ' ', array([ -3.35937500e-01,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
(112, ' ', array([ -1.00000000e+00,  -3.05175781e-05,   1.00000000e+00,
         0.00000000e+00,   0.00000000e+00]))
They look good to me.

**Comment by xaviervasques** [comment]
Extract of "data" "sample" and model_weights.h5
https://www.dropbox.com/s/j15imcnk13duju4/model_weights.h5?dl=0

**Comment by kevinhughes27** [comment]
Model weights are hard to interpret - instead can you run the data set visualizer and upload a gif or video?

**Comment by bzier** [comment]
Sorry guys, I've been away from this project for a bit. It sounds like this is mostly an issue with recording the inputs from the controller and/or training. If you determine that there is an issue with the play/execution in the environment, please let me know and I'll do my best to look into it.

**Comment by xaviervasques** [comment]
Thanks Kevin, you will find  in dropbox the data set visualizer in one race: https://www.dropbox.com/sh/upkul266fjk4bfs/AADVdtnnmvFfDSnL9vwv0Qina?dl=0

**Comment by xaviervasques** [comment]
Just to add, I modified within the code: 

In utils.py
class Screenshot(object):
    SRC_W = 640
    SRC_H = 480
    SRC_D = 3

    OFFSET_X = 0
    OFFSET_Y = 0
I replaced SRC by IMG:

class Screenshot(object):
    IMG_W = 640
    IMG_H = 480
    IMG_D = 3

    OFFSET_X = 0
    OFFSET_Y = 0

because of a small error. 

In gymplus, in config.yml

USE_XVFB: false instead of "true"

**Comment by kevinhughes27** [comment]
Please PR any config issues that are our fault! I don't often do a full end to end test of all new contributions so these things come up from time to time.

Can you plot that data for me or use the visualisation tool? Its not obvious to see just from a csv file.

**Comment by bzier** [comment]
@kevinhughes27 The .mov files must not have finished uploading right away. When I originally went to the dropbox link, there was only the .csv and .h5 files. Now there are two .mov recordings as well.

**Comment by kevinhughes27** [comment]
Oh sweet!

**Comment by kevinhughes27** [comment]
I don't think you are driving 'smooth enough' and that is your issue. 

![image](https://user-images.githubusercontent.com/1965489/29490137-6ef4faa0-8501-11e7-8d0a-281f922c25c8.png)

I wrote about this aliasing issue in the accompanying blog post: https://kevinhughes.ca/blog/tensor-kart. TensorKart is pretty vulnerable to a 'local minima' where the AI learns to never turn over the test set.

**Comment by xaviervasques** [comment]
Hi Kevin, I tried to drive smoothly (see "record_mario_3" in https://www.dropbox.com/sh/upkul266fjk4bfs/AADVdtnnmvFfDSnL9vwv0Qina?dl=0) but it doesn't work for me ...  I also tried different ways to drive. Can you share your samples to check if it is really the way I drive  ? Thank you

**Comment by kevinhughes27** [comment]
Did you read my blog post? I can still see examples of aliasing in your turns. Its problematic because of the ratio of turns to straight in the dataset you are training on. I'll try and upload a video of my data later tonight.

**Comment by kevinhughes27** [comment]
Here are some screenshots from my dataset:

mid turn:
![selection_001](https://user-images.githubusercontent.com/1965489/29593339-b0eb493c-8777-11e7-80b3-cfc8abc71fd7.png)

after a large turn:
![selection_002](https://user-images.githubusercontent.com/1965489/29593348-bd3340fa-8777-11e7-9996-1e44e92c8f31.png)

You'll see I'm not perfectly smooth either but I always try and never hit full zero during a turn. You need enough good examples of turning in your data or else there isn't enough for the network to learn.

**Comment by xaviervasques** [comment]
Hi Kevin and thank you for your replies. I tried many ways to drive and smoother, it is a little bit better but not enough to race a complete lap without issues :) It turns but not really as expected because sometimes it takes the wrong direction (for exemple Mario turn left and he should turn right or it doesn't turn enough or he still goes straigth ahead). Could you share with me a data set that works perfectly for you ?

**Comment by kevinhughes27** [comment]
The point of TensorKart is not the final result - its to help people understand the process. That is why I didn't publish my data set. The AI trained from my data set could complete Luigi Raceway except for entering the tunnel and generalised to a section of un-trained track from Royal Raceway. I wouldn't call my result perfect.

What I am saying is that it sounds like your results aren't super far off from mine. What do you hope to achieve by using the exact same data as me?

**Comment by xaviervasques** [comment]
Hi Kevin, I understand. So it sounds like we have the same pretty results. I will close this one and thanks for your help !
